Src/beebwin.cpp

```diff
@@ -1248,13 +1294,179 @@ void BeebWin::ScaleJoystick(unsigned int x, unsigned int y)
+	{
+		if (joyInfoEx.dwPOV >= 29250 || joyInfoEx.dwPOV < 6750)
+			axes |= (1 << JOYSTICK_AXIS_HAT_UP);
+		if (joyInfoEx.dwPOV >= 2250 && joyInfoEx.dwPOV < 15750)
+			axes |= (1 << JOYSTICK_AXIS_HAT_RIGHT);
+		if (joyInfoEx.dwPOV >= 11250 && joyInfoEx.dwPOV < 24750)
+			axes |= (1 << JOYSTICK_AXIS_HAT_DOWN);
+		if (joyInfoEx.dwPOV >= 20250 && joyInfoEx.dwPOV < 33750)
+			axes |= (1 << JOYSTICK_AXIS_HAT_LEFT);
+	}
+
+	return axes;
+}
+
+/****************************************************************************/
+void BeebWin::TranslateOrSendKey(int vkey, bool keyUp)
+{
+	int row, col;
+	if (m_JoystickTarget == NULL)
+	{
+		TranslateKey(vkey, keyUp, row, col);
+	}
+	else if (!keyUp)
+	{
+		// Keyboard input dialog is visible - translate button down to key down message and send to dialog
+		PostMessage(m_JoystickTarget, WM_KEYDOWN, vkey, 0);
+	}
+}
+
+/****************************************************************************/
+void BeebWin::TranslateJoystickMove(int joyId, JOYINFOEX& joyInfoEx)
+{
+	const int axNum = JOYSTICK_AXES_COUNT;
+
+	JOYCAPS& caps = (joyId == 1) ? m_Joystick2Caps : m_JoystickCaps;
+	unsigned int deadband = (joyId == 1) ? m_Joystick2Deadband : m_Joystick1Deadband;
+	int& prevAxes = (joyId == 1) ? m_Joystick2PrevAxes : m_Joystick1PrevAxes;
+	int vkeys = (joyId == 1) ? BEEB_VKEY_JOY2_AXES : BEEB_VKEY_JOY1_AXES;
+
+	unsigned int axes = mainWin->GetJoystickAxes(caps, deadband, joyInfoEx);
+
+	if (axes != prevAxes)
+	{
+		for (int axId = 0; axId < axNum; ++axId)
+		{
+			if ((axes & ~prevAxes) & (1 << axId))
+			{
+				TranslateOrSendKey(vkeys + axId, false);
+			}
+			else if ((~axes & prevAxes) & (1 << axId))
+			{
+				TranslateOrSendKey(vkeys + axId, true);
+			}
+		}
+		prevAxes = axes;
+	}
+}
+
+/****************************************************************************/
+void BeebWin::TranslateJoystickButtons(int joyId, unsigned int buttons)
+{
+	const int btnNum = 32;
+
+	int& prevBtns = (joyId == 1) ? m_Joystick2PrevBtns : m_Joystick1PrevBtns;
+	int vkeys = (joyId == 1) ? BEEB_VKEY_JOY2_BTN1 : BEEB_VKEY_JOY1_BTN1;
+
+	if (buttons != prevBtns)
+	{
+		for (int btnId = 0; btnId < btnNum; ++btnId)
+		{
+			if ((buttons & ~prevBtns) & (1 << btnId))
+			{
+				TranslateOrSendKey(vkeys + btnId, false);
+			}
+			else if ((~buttons & prevBtns) & (1 << btnId))
+			{
+				TranslateOrSendKey(vkeys + btnId, true);
+			}
+		}
+		prevBtns = buttons;
+	}
+}
+
+/****************************************************************************/
+void BeebWin::TranslateJoystick(int joyId)
+{
+	if (mainWin->m_JoystickToKeys)
+	{
+		JOYINFOEX joyInfoEx;
+		joyInfoEx.dwSize = sizeof(joyInfoEx);
+		joyInfoEx.dwFlags = JOY_RETURNALL | JOY_RETURNPOVCTS;
+		if (!joyGetPosEx(joyId, &joyInfoEx))
+		{
+			mainWin->TranslateJoystickMove(joyId, joyInfoEx);
+			mainWin->TranslateJoystickButtons(joyId, joyInfoEx.dwButtons);
+		}
+	}
+}
+
 /****************************************************************************/
 void BeebWin::ResetJoystick(void)
```

Comment:
Should we remove this function, if it doesn't do anything?

Reply:
Yes